src/rust/catloop/mod.rs

```diff
@@ -264,19 +305,15 @@ impl CatloopLibOS {
         trace!("close() qd={:?}", qd);
 
         // Remove socket from sockets table.
-        match self.sockets.remove(&qd) {
+        match self.qtable.free(&qd) {
             // Socket is not bound to a duplex pipe.
-            Some(None) => {
-                self.qtable.free(qd);
-                Ok(())
-            },
-
-            // Socket is bound to a duplex pipe.
-            Some(Some((_, duplex_pipe))) => {
-                self.listening.remove(&qd);
-                duplex_pipe.close()?;
-                self.qtable.free(qd);
-                Ok(())
+            Some(queue) => match queue.get_pipe() {
+                Some(duplex_pipe) => {
+                    self.listening.remove(&qd);
+                    duplex_pipe.close()?;
+                    Ok(())
```

Comment:
We are now releasing the queue descriptor even if we fail to release the underlying resources.

Reply:
Fixed this.